### ISSUE TITLE
Build sqlitebrowser from source, sqlcipher support (Windows)

### ISSUE THREAD

**Issue by huy-cong** [sqlitebrowser/sqlitebrowser#482]
Greetings everyone,
My goal is to re-compile sqlitebrowser for Windows with the support to open encrypted database. I downloaded the source code and already "scan" the support you insert in source (#ifdef ENABLE_SQLCIPHER)
I have Qt 5.4 & sqlcipher (binary & almagation files - license trial from Zetetic, and source code avaible ofc), but I dont know which steps to follow to attach sqlcipher with sqlitebrowser.
For now I did:
qmake CONFIG+=sqlcipher
make

But it gives me this error (which I though do not related much on sqlcipher...):

"
make[2]: Entering directory '/d/LibExternal/sqlitebrowser-3.8.0/libs/antlr-2.7.7'
g++ -c -pipe -fno-keep-inline-dllexport -O2 -frtti -Wall -Wextra -fexceptions -mthreads -DUNICODE -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_CORE_LIB -I'.' -I'.' -I'../../../../Tools/local32/dst/qt/include' -I'../../../../Tools/local32/dst/qt/include/QtGui' -I'../../../../Tools/local32/dst/qt/include/QtCore' -I'release' -I'../../../../Tools/local32/dst/qt/mkspecs/win32-g++'  -o release/NoViableAltForCharException.o src/NoViableAltForCharException.cpp
In file included from D:/Tools/mingw32/i686-w64-mingw32/include/_mingw.h:13:0,
                 from D:/Tools/mingw32/i686-w64-mingw32/include/crtdefs.h:10,
                 from D:/Tools/mingw32/i686-w64-mingw32/include/wchar.h:9,
                 from D:/Tools/mingw32/i686-w64-mingw32/include/c++/cwchar:44,
                 from D:/Tools/mingw32/i686-w64-mingw32/include/c++/bits/postypes.h:40,
                 from D:/Tools/mingw32/i686-w64-mingw32/include/c++/bits/char_traits.h:40,
                 from D:/Tools/mingw32/i686-w64-mingw32/include/c++/string:40,
                 from ./antlr/ANTLRException.hpp:12,
                 from ./antlr/RecognitionException.hpp:12,
                 from ./antlr/NoViableAltForCharException.hpp:12,
                 from src/NoViableAltForCharException.cpp:8:
D:/Tools/mingw32/i686-w64-mingw32/include/sec_api/string_s.h: In function 'errno_t antlr::strcpy_s(char (&)[__size], const char*)':
D:/Tools/mingw32/i686-w64-mingw32/include/sec_api/string_s.h:28:3: error: '::strcpy_s' has not been declared
   __CRT_SECURE_CPP_OVERLOAD_SECURE_NAMES_0_2_(errno_t,strcpy_s,char,,_Dst,size_t,,_SizeInBytes,const char *,,_Src)
   ^
D:/Tools/mingw32/i686-w64-mingw32/include/sec_api/string_s.h:28:3: note: suggested alternative:
In file included from D:/Tools/mingw32/i686-w64-mingw32/include/_mingw.h:13:0,
                 from D:/Tools/mingw32/i686-w64-mingw32/include/crtdefs.h:10,
                 from D:/Tools/mingw32/i686-w64-mingw32/include/wchar.h:9,
                 from D:/Tools/mingw32/i686-w64-mingw32/include/c++/cwchar:44,
                 from D:/Tools/mingw32/i686-w64-mingw32/include/c++/bits/postypes.h:40,
                 from D:/Tools/mingw32/i686-w64-mingw32/include/c++/bits/char_traits.h:40,
                 from D:/Tools/mingw32/i686-w64-mingw32/include/c++/string:40,
                 from ./antlr/ANTLRException.hpp:12,
                 from ./antlr/RecognitionException.hpp:12,
                 from ./antlr/NoViableAltForCharException.hpp:12,
                 from src/NoViableAltForCharException.cpp:8:
D:/Tools/mingw32/i686-w64-mingw32/include/sec_api/string_s.h:28:3: note:   'antlr::strcpy_s'
   __CRT_SECURE_CPP_OVERLOAD_SECURE_NAMES_0_2_(errno_t,strcpy_s,char,,_Dst,size_t,,_SizeInBytes,const char *,,_Src)
   ^
"

Do you have any ideas what is gone wrong? Is there anything I miss in the configuration steps?

Thanks a lot for any helps,
Best regards,

Huy Cong Vu

**Comment by huy-cong** [comment]
Hi guys, 
Sorry for duplicate posts in here and your private mail @justinclift, I guess we should discuss here from now on, It'll be easier for both of us :)

**Comment by justinclift** [comment]
Interesting questions.  @rp- creates our Windows release builds using Visual Studio (I think, that's from memory).  Our nightly Windows builds use an older process, cross-compiling from Linux using [MXE](https://github.com/sqlitebrowser/sqlitebrowser/blob/master/BUILDING.md#cross-compiling-for-windows).

From your error output above, it looks like you're using mingw32.  I'm unsure if anyone's tried that yet. :grin: 

And yeah, lets discuss here.  It distributes the problem solver / knowledge wider. :wink:

**Comment by huy-cong** [comment]
Yes indeed, I usually compile my other Qt projects with mingw32 and this configuration and it works fine. 
This error showned up when I tried to compile antlr lib. I'm not sure that it depends on string_s.h (which is standard) from other packages or not.

**Comment by justinclift** [comment]
Yeah, I'm not sure either.  I don't code C++ (at this point), I more help out this project with compiling on OSX and making packages there.

Is there any way you can add an #include (or C++ equivalent) or something for string_s.h, to see if that works?  Note - That's just me guessing. :grin:

**Comment by glawrence** [comment]
I keep meaning to look Windows builds, so will try and get restarted on this

**Comment by rp-** [comment]
building sqlitebrowser itself is not much of a problem with cmake on windows.

Its more a problem of getting the dependencies correctly build, e.g. sqlite is a bit of a pitty because there are no build files for VS and you have to do them on your own with the correct .def file.

I don't build with qmake anymore and don't know how a mingw build will work out on windows.

**Comment by huy-cong** [comment]
I'll try with cmake then. 
Following this link, I must have MXE to compile? I had msys2 already https://github.com/sqlitebrowser/sqlitebrowser/blob/master/BUILDING.md#cross-compiling-for-windows

And then in the cmake file, I find some absolute path for the libs, which means that I have to set it to the right path on my system?
"
if(WIN32 AND MSVC)
	if(CMAKE_CL_64)
		set(QT5_PATH "E:/Qt64/Qt5.5.1/5.5/msvc2013_64")
		set(SQLITE3_PATH "E:/libs/sqlite3_x64")
		set(OPENSSL_PATH "E:/libs/openssl-1.0.2a-x64_86-win64")
		set(VSREDIST "vcredist_x64.exe")
	else()
		set(QT5_PATH "E:/Qt/Qt5.5.1/5.5/msvc2013")
		set(SQLITE3_PATH "E:/libs/sqlite3")
		set(OPENSSL_PATH "E:/libs/openssl-1.0.2a-i386-win32")
		set(VSREDIST "vcredist_x86.exe")
	endif()
	set(USE_QT5 TRUE)
	set(CMAKE_PREFIX_PATH "${QT5_PATH};${SQLITE3_PATH}")
endif()
"

**Comment by rp-** [comment]
No, MXE is for cross compiling from linux (for our nightly builds)

The installer .exe is created using VS2013 and cmake.
Yes you would need to set the paths to our install/library directories

**Comment by huy-cong** [comment]
Oh wrong call, it is a cross-compiling for Windows from Linux, won't work with me then :(
So if I stay in Windows and use cmake (with mingw) is it still possible?

**Comment by rp-** [comment]
yes it should be possible

**Comment by huy-cong** [comment]
I'm trying to build with MingW 32 bits without success for now. 
When I ran cmake -Dsqlcipher, it gives me this error:

-- Could NOT find Antlr2 (missing:  ANTLR2_LIBRARIES ANTLR2_INCLUDE_DIRS)
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
LIBSQLITE

Eventhough I already add the path to my sqlcipher lib files:
	
find_library(LIBSQLITE ${LIBSQLITE_NAME} HINTS /local32/dst/sqlcipher/lib)
set(ADDITIONAL_INCLUDE_PATHS ${ADDITIONAL_INCLUDE_PATHS} /local32/dst/sqlcipher/include)

I had libsqlcipher.a & libsqlcipher.la in lib directory (build from sqlcipher source with same MingW 32 tools). I wonder is these 2 files the CMake are looking for?

2nd questions: I must set these variables to the path of sources or binary?
SQLITE3_PATH 
OPENSSL_PATH

**Comment by rp-** [comment]
those paths should point to foremost binaries
a directory structure like SQLITE3_PATH/lib, SQLITE3_PATH/include

this is how cmake will search for the includes and lib files

**Comment by rp-** [comment]
see also:
https://cmake.org/cmake/help/v3.3/variable/CMAKE_PREFIX_PATH.html

**Comment by huy-cong** [comment]
I bypass the error when run "cmake -Dsqlcipher=1" when force link library to a pre-compiled sqlite3.dll. However, when I run "make", the same error that I had in the 1st post with qmake is reproduced, and I still have no idea why...

**Comment by rp-** [comment]
It looks like a define mismatch, someone(mingw or antlr) wants to call strcpy_s, but this function is only avialable I think on MS visual c compiler, it is some pseudo secure variant of strcpy.

http://sourceforge.net/p/mingw-w64/mailman/message/25751607/

**Comment by huy-cong** [comment]
Finally, I manage to compile with MSVC compiler & Visual studio + sqlcipher lib precompiled, it seems that it won't work without MSVC compiler. I now have sqlcipher browser works perfectly.
Thanks a lot for all your helps! :)

**Comment by glawrence** [comment]
Nice work @huy-cong that's great news. I have been working on something similar with a view to documenting the steps on our Wiki but I am behind you in terms of progress! Would you have time to write up your steps and anything else you noted along the way? Please. If you just sketch something out roughly in a reply here I can then pick this up and format for the wiki, adding my own experience. We can then also update other documentation, however I wanted to make your contribution easy for you! Thanks.

**Comment by huy-cong** [comment]
Hi @glawrence 

I would love to contribute for your community. But I don't know what to write much about, it's pretty straightforward in fact. I just hesitated to choose the build tools (qmake or cmake, gcc or MSVC...)

So my build set is : MSVC + cmake, in Microsoft Visual Studio 2013

 In general you guys all setup already all .vcproj file, config & build step. All that I have to modify is the paths to dependancies (sqlcipher, Qt for MSVC 32 bits) in CMakeLists.txt file (which for now is not generic, but point to local path of your build, I think for the future you can setup some find[Package] with CMake for those dependancies). 

The roughest part must be figuring out which sqlcipher .dll I have to link with, because I have differents builds from differents sources, but I choosed a C/C++ DLL for Windows pre-compiled that I order a trial from SQLCipher main site web.

Rather than make the package generic, I just modify it to make it work on my local build. I don't know if it's useful to improve your work?

**Comment by glawrence** [comment]
Thanks @huy-cong that's helpful and appreciated. I will close this issue down now as I believe we are all done.

**Comment by justinclift** [comment]
@huy-cong Do you still have your CMakeLists.txt file around?

I'm trying to get this working with MSVC 2013 + CMake ([documenting the progress here](https://github.com/sqlitebrowser/sqlitebrowser/wiki/Notes-for-compiling-on-Windows)), but am hitting the ANTLR same error you mention above.

Any ideas?

**Comment by justinclift** [comment]
Managed to get past that error, and get CMake to generate the project files.  It seems like the project is being compiled in x86 mode though, not x64, as it then complains about a mismatch.

Does anyone know where to set that?  I was **very** careful to only download x64 version of Qt, OpenSSL, etc, and I'm running Win 8.1 x64.  So, I'm unsure where it's getting x86 from. ???

**Comment by drescherjm** [comment]
> Does anyone know where to set that?

You set this in cmake-gui by selecting a 64 bit generator.

**Comment by justinclift** [comment]
Thanks @drescherjm I'll take a look at that again.  I was trying things from the command line, as the GUI didn't seem to have a way to "Open File" to select an existing project.

I'll take a look again. :smile:

**Comment by justinclift** [comment]
Hmmm, that seems to have completed the build.  Ok, this might be worth putting more time into after all.

Thanks heaps @drescherjm, that seems to have solved that problem. :grin:

Now to see if I can get the "package" step working...

**Comment by justinclift** [comment]
Holy crap.  It worked.

![first compile of db4s on windows](https://cloud.githubusercontent.com/assets/406299/14868193/506c330c-0cc4-11e6-8bff-3e1a30df4d5e.png)

Tomorrow, I'm going to document the living stuffing out of the process to make that work. :smile:

**Comment by drescherjm** [comment]
BTW, You can also set the generator on the command line via cmake -G "generator" 

cmake --help 

will list the options.

**Comment by justinclift** [comment]
Awesome, that's really, really helpful. :smile:

**Comment by justinclift** [comment]
The target for building a package with NSIS is also working too.  Hmmm, it might be possible for us to get SQLCipher support working too.  Another thing to try tomorrow. :smile:

**Comment by justinclift** [comment]
Update - I've gone through the process (from a newly installed Win 8.1 desktop) to compiled & packages 64-bit version on DB4S on Windows, taking 104 screenshots along the way.

Mentally brain fogged now, so the cropping and turning-them-into-useful-docs process is going to have to wait a few hours before starting, or perhaps tomorrow morning.

**Comment by justinclift** [comment]
For anyone interested, the docs-in-development for compiling DB4S on Windows are here:

&nbsp; &nbsp; https://github.com/justinclift/sqlitebrowser/wiki/Setting-up-a-Win64-development-environment-for-DB4S

They're in an ok state now.  Some fine tuning still needs to be done to make things prettier, but the content is workable as-is. :smile:

**Comment by justinclift** [comment]
SQLCipher seems to compile ok, and DB4S seems to pretty much "just work" with it when the right includes/directory-paths/filenames are given to it.  Initial package here for testing, if anyone wants to give it a shot:

&nbsp; &nbsp; http://nightlies.sqlitebrowser.org/win64/sqlitebrowser-win64-with_sqlcipher-201605041823.exe

**Comment by kiwina** [comment]
just tried no luck

**Comment by justinclift** [comment]
@kiwina Heh Heh Heh, that's a bit light on details. :wink:

Which bits did you try, and what went wrong?  Screenshots of error messages and similar are helpful, so we can investigate and try to fix it.

As an extra thought, I've been putting time into some further instructions too, for compiling SQLCipher on Windows:

&nbsp; &nbsp; https://github.com/justinclift/sqlitebrowser/wiki/Win64-setup-—-Compiling-SQLCipher

They're also still a "work in progress", but might be useful.

Anyway, if you're ok to let us know what broke when you tried things, that would help. :smile:

**Comment by iKlsR** [comment]
For simple build purposes, the compile step for using MSVC is way too long with so much stuff to dl. I'm not sure if there were any large changes between this post and now but I got the browser to build on windows by just opening the `.pro` in Qt Creator.

Only thing is I had to build the amalgamation as a static and a shared lib and plob those in Qt's mingw folder since the solution calls for `-lsqlite`. Have no need of the cipher atm but I think it would be fairly trivial to add as well or maybe I went about this the wrong way or this was never an issue. Also if anyone chances on this, I ran into a small newb quirk, I didn't follow the naming convention that I built with and renamed my shared library to `sqlite3.dll`, it has to be the the same name as the `.def` file or it will be unusable, I had to use PE Explorer to check library dependencies, wasted a good few minutes there.

Another gotcha is that I think I need a cert for network stuffs but for basic playing around one should be fine.
